### ISSUE TITLE
Document how to "determined the impact on package closure size"

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#51368]
## Issue description

The PR template says: "(by running nix path-info -S before and after)"

I think that's not very clear. I have created a one-liner that directly outputs the result as text.

Should i create a bash script that expects the old and new path as parameter and outputs the result that you can put directly in your PR?

We can also explain it in more detail on https://nixos.org/nixpkgs/manual/#chap-submitting-changes

### Steps to reproduce

I have mumble installed from 18.09 channel (it's broken) and create a PR to fix it.

```
$ BEFORE=$(nix path-info -S /run/current-system/sw/bin/mumble | cut -d' ' -f3); AFTER=$(nix path-info -S /nix/store/ynq5fwv8n9m5vb2qy5zfl1dn164dhiw8-mumble-1.2.19 | cut -d' ' -f3); PERCENT=$(nix run nixpkgs.bc -c bc <<< "scale=2;$AFTER * 100 / $BEFORE - 100"); DIFF=$(nix run nixpkgs.bc -c bc <<< "scale=2;($AFTER - $BEFORE) / 1000000"); echo "Closure size change $PERCENT% ($DIFF MB)."
Closure size change 0% (0 MB).
```

Other output examples:

- `Closure size change 2.56% (9.09 MB)`
- `Closure size change -2.48% (-8.99 MB)`

The second path is from my fix.
```
[davidak@X230:~]$ nix path-info -S /run/current-system/sw/bin/mumble | cut -d' ' -f3
363869624

[davidak@X230:~]$ nix path-info -S /nix/store/ynq5fwv8n9m5vb2qy5zfl1dn164dhiw8-mumble-1.2.19 | cut -d' ' -f3
363869672
```

**Comment by dtzWill** [comment]
This is definitely something we should have tooling for, at least to facilitate these comparisons!

Hydra graphs package closures over time, FWIW O:).

I think this bullet might be intended to also include a more complicated notion of "does this update push some huge dependency into downstream packages that don't need it" -- which maybe tooling can help investigate as well? But that's a bit trickier.

**Comment by davidak** [comment]
I made a script. Where is the right place for it, so Nix(OS) users can use it?

Maybe a nixpkgs package and command using `nix run`?

```
#! /usr/bin/env nix-shell
#! nix-shell -i bash -p bc

# assign variables
PATH1=$1
PATH2=$2

NAME="compare-path-size"
VERSION="0.1.0"

# declare functions
function print_usage() {
  echo -e "USAGE: $NAME PATH1 PATH2\n"
  echo "Compare closure size of 2 paths from nix store."
  exit
}

function print_version() {
  echo "$NAME version $VERSION"
  exit
}

function check_path() {
  nix path-info -S $1 &>/dev/null
  exit_code=$?
  if [ $exit_code -ne 0 ]; then
    echo "Error: can't get closure size of $1" >&2
    exit $exit_code
  fi
}

function get_path_size() {
  local OUTPUT="$(nix path-info -S $1)"
  # return part after last ' ' (size)
  echo ${OUTPUT##* }
}

# check for parameters
case "$1" in
  -h|--help)
    print_usage
    ;;
  --version)
    print_version
    ;;
esac

# check if 2 arguments passed
if ! [ $# -eq 2 ]; then
  print_usage
fi

# check if input is a valid path
check_path $PATH1
check_path $PATH2

# get path size in bytes
PATH1_SIZE=$(get_path_size $PATH1)
PATH2_SIZE=$(get_path_size $PATH2)

# calculate difference
PERCENT=$(echo "scale=2;$PATH2_SIZE * 100 / $PATH1_SIZE - 100" | bc)
MB=$(echo "scale=2;($PATH2_SIZE - $PATH1_SIZE) / 1000000" | bc)

# output difference
echo "Closure size difference is $MB MB ($PERCENT%)."
```

**Comment by JohnAZoidberg** [comment]
The template should at least mention that one should run it on the nix-store path.
I tried running it on the `result` symlink and that just gave me errors that didn't lead me to the correct conclusions.

**Comment by jorsn** [comment]
> I tried running it on the `result` symlink and that just gave me errors that didn't lead me to the correct conclusions.

The argument must be an installable. If you call it `result`, nix assumes it is a nixpkgs attribute and does not find it. If you call it `./result`, nixpkgs handles it as a path, thus correctly.

**Comment by fzakaria** [comment]
@davidak you might like this shell function which I have in my home-manager zshrc

```bash
# Figure out the closure size for a certain package
# ex. nix-closure-size $(which exa)
nix-closure-size() {
  nix-store -q --size $(nix-store -qR $(readlink -e $1) ) | \
  awk '{ a+=$1 } END { print a }' | \
  /nix/store/cc64vcf0g3l5ck0c5pfwflcacb8x7i4d-coreutils-8.31/bin/numfmt --to=iec-i
}
```

```bash
❯ nix-closure-size $(which exa)
34Mi
```

**Comment by FliegendeWurst** [comment]
This line was removed from the PR template by now (#117537), but it would still be a good idea to document this somewhere..

**Comment by c0bw3b** [comment]
Indeed, I never know for sure what to use between `nix path-info -s` (lowercase), `nix path-info -S` (uppercase) and `nix-store -q --size`.

Plus how to get human-readable numbers because without `-h` nix path-info outputs raw numbers that you _assume_ to be bytes but it's not clear in the output.

**Comment by jorsn** [comment]
Document what exactly? For me (flakes activated) `nix path-info --help` gives me a nice, colorful man page where the options are described. I'd call this documented. There, I also see that you can type `nix path-info --closure-size --human-readable`. Is this rememberable?

**Comment by davidak** [comment]
>Document what exactly?

Document how to "determined the impact on package closure size"

If that is still relevant to know.

**Comment by jorsn** [comment]
Yes, that's the issue title. Maybe I could have phrased it better: Since this is no longer requiredd for PRs, do @FliegendeWurst and @c0bw3b notice any particular lack of documentation? I mean, a script for comparing on top of determining the sizes is nice, but is it as important as to be officially endorsed, since its functionality no longer officially required?
